Here it is. I hope it's useful

List of diff/changes

LOGIN_NOCOOKIE support for cookie only login
838,861 support for logging IP
978,1003 date patch sten.eriksson@udac.se Thu, 13 Jul 2000 11:33:26
4791,4799 date patch sten.eriksson@udac.se Thu, 13 Jul 2000 11:33:26
1040,1057 allow more text to be checked for links
1231,1238 rfc822 quoting
1848,1856 fix for addresses where leading newline appears
1920,1928 more ip logging
3560,3573 ldap enhancement Niklas Edmundsson <nikke@ing.umu.se> Thu, 15 Jun 2000 21:29:13
6117,6171 ldap enhancement Niklas Edmundsson <nikke@ing.umu.se> Thu, 15 Jun 2000 21:29:13
3845,3854 timezone fix from imho list
3975,3991 only use rfc1522 encoding if date is really 8 bit
5086,5139 fix to handle pine 4 address books
5189,5243 import simeon export format
6252,6258 (and others) fix cookie support for roxen 2.0
7531,7537 (and others) fix cookie support for roxen 2.0
7228,7253 fix to stop pike error if user clicks move to mailbox without selecting
7289,7301 next message selected after `trashthis' now sort order dependent
7434,7464 show all hdrs with forward if all hdrs set to appear
8083,8096 show link that returns to compose if ldap fails to find any matches
8299,8313 show next block now sort order dependent
8521,8531 show next/prev message now sort order dependent
8565,8575 full hdrs patch (not sure if this works with old pike) stewa@lysator.liu.se Sun Jul  2 20:41:16 2000
8958,8985 support for message/rfc822 inline
9021,9033 allow plain/text to be downloaded
9835,9843 extra text to help get out of search


-- 
Alan Thew
alan.thew@liverpool.ac.uk    http://www.pobox.com/~alan.thew

---------- Forwarded message ----------

*** imho.orig	Thu Sep 14 21:33:58 2000
--- imho.pike.dev	Sun Oct 29 20:30:09 2000
***************
*** 44,49 ****
--- 44,55 ----
      Xavier Beaudouin <xbeaudouin@isdnet.net> 2000
    IMAP from LDAP by
      Lai Yiu Fai <ccyflai@ust.hk> 2000
+   UoL changes:
+   full hdrs patch by stewa@lysator.liu.se Sun Jul  2 20:41:16 2000
+   extra date formats sten.eriksson@udac.se Thu, 13 Jul 2000 11:33:26
+   ldap improvements Niklas Edmundsson <nikke@ing.umu.se> Thu, 15 Jun 2000 21:29:13
+   Jake Gannon all HTML type fixes
+   Alan Thew all other pike fixes/enhancements
  */


***************
*** 57,62 ****
--- 63,70 ----
  #define LOGINPROMPT -1
  #define LOGINFAILED -2
  #define LOGINFAILEDIMAP -3
+ //AJT
+ #define LOGIN_NOCOOKIE -4

  #define MAINMENU 3
  #define LOGOUT 4
***************
*** 135,140 ****
--- 143,149 ----
  #define M_SAVEDUSERINTERFACE 12
  #define M_USERINTERFACE 13
  #define M_LOGIN_OK 14
+ #define M_NOCOOKIE 4 //AJT
  // page header
  #define M_NEWMAIL 50
  #define M_MAILBOX 51
***************
*** 330,335 ****
--- 339,345 ----
  constant screennames = ([ LOGINPROMPT : ({ "login", -1 }),
  			 LOGINFAILED : ({ "login", -1 }),
  			 LOGINFAILEDIMAP : ({ "login", -1 }),
+ 			 LOGIN_NOCOOKIE : ({ "login", -1 }),
  			 LOGOUT : ({ "logout", M_LOGGEDOUT }),
  			SETUP : ({ "setup", M_PREFS }),
  			COMPOSE : ({ "compose", M_COMPOSEMAIL }),
***************
*** 402,409 ****
  "    <if not defined=\"imhonewmail is 0\"><comment>No new mail</comment>\n" \
  "     <body>\n" \
  "     <center>\n" \
! "     <table border=\"0\"><tr><td bgcolor=\"#000070\">\n" \
! "      <imho_image image=\"newmail\" quant=\"250\" insideenvcolor=\"#202020\" bg=\"#000070\" scale=\"0.5\" /><br />\n" \
  "     </td></tr>\n" \
  "     <if supports=\"javascript\">\n" \
  "      <script>\n" \
--- 412,419 ----
  "    <if not defined=\"imhonewmail is 0\"><comment>No new mail</comment>\n" \
  "     <body>\n" \
  "     <center>\n" \
! "     <table border=\"0\"><tr><td bgcolor=\"#3E60AA\">\n" \
! "      <imho_image image=\"newmail\" quant=\"250\" insideenvcolor=\"#202020\" bg=\"#3E60AA\" scale=\"0.5\" /><br />\n" \
  "     </td></tr>\n" \
  "     <if supports=\"javascript\">\n" \
  "      <script>\n" \
***************
*** 428,435 ****
  "    <else>\n" \
  "     <body>\n" \
  "     <center>\n" \
! "     <table bgcolor=\"#000070\" border=\"0\"><tr><td>\n" \
! "      <imho_image image=\"nonewmail\" quant=\"250\" insideenvcolor=\"#202020\" bg=\"#000070\" scale=\"0.5\" />\n" \
  "     </td></tr></table>\n" \
  "     </center>\n" \
  "     </body>\n" \
--- 438,445 ----
  "    <else>\n" \
  "     <body>\n" \
  "     <center>\n" \
! "     <table bgcolor=\"#3E60AA\" border=\"0\"><tr><td>\n" \
! "      <imho_image image=\"nonewmail\" quant=\"250\" insideenvcolor=\"#202020\" bg=\"#3E60AA\" scale=\"0.5\" />\n" \
  "     </td></tr></table>\n" \
  "     </center>\n" \
  "     </body>\n" \
***************
*** 460,473 ****
  "       <body text=\"black\" bgcolor=\"#dde1f0\" link=\"black\" vlink=\"black\" alink=\"black\">\n" \
  "       <cset preparse=\"\" scope=\"var\" variable=\"banner\"><imho_banner /></cset>\n" \
  "       <if match=\"&roxen.version; is Roxen*\">\n" \
! "        <imho_image image=\"bannerleft\" text=\"&var.banner;\" quant=\"250\" clouds=\"white\" fg=\"white\" bg=\"#000070\" align=\"left\" border=\"0\" hspace=\"0\" />\n" \
  "       </if>\n" \
  "       <else>\n" \
  "       <formoutput quote=\"$\">\n" \
! "        <imho_image image=\"bannerleft\" text=\"$banner$\" quant=\"250\" clouds=\"white\" fg=\"white\" bg=\"#000070\" align=\"left\" border=\"0\" hspace=\"0\" />\n" \
  "       </formoutput>\n" \
  "       </else>\n" \
! "       <imho_image image=\"bannerright\" bg=\"#000070\" width=\"100%\" border=\"0\" hspace=\"0\" />\n" \
  "       </body>\n" \
  "      </if><comment>End banner frame</comment>\n" \
  "\n" \
--- 470,483 ----
  "       <body text=\"black\" bgcolor=\"#dde1f0\" link=\"black\" vlink=\"black\" alink=\"black\">\n" \
  "       <cset preparse=\"\" scope=\"var\" variable=\"banner\"><imho_banner /></cset>\n" \
  "       <if match=\"&roxen.version; is Roxen*\">\n" \
! "        <imho_image image=\"bannerleft\" text=\"&var.banner;\" quant=\"250\" clouds=\"white\" fg=\"white\" bg=\"#3E60AA\" align=\"left\" border=\"0\" hspace=\"0\" />\n" \
  "       </if>\n" \
  "       <else>\n" \
  "       <formoutput quote=\"$\">\n" \
! "        <imho_image image=\"bannerleft\" text=\"$banner$\" quant=\"250\" clouds=\"white\" fg=\"white\" bg=\"#3E60AA\" align=\"left\" border=\"0\" hspace=\"0\" />\n" \
  "       </formoutput>\n" \
  "       </else>\n" \
! "       <imho_image image=\"bannerright\" bg=\"#3E60AA\" width=\"100%\" border=\"0\" hspace=\"0\" />\n" \
  "       </body>\n" \
  "      </if><comment>End banner frame</comment>\n" \
  "\n" \
***************
*** 488,494 ****
  "         <if defined=\"imhoscreen is mailindex\">\n" \
  "          <h3><imho_mailbox />:</h3>\n" \
  "         </if>\n" \
! "         <imho_main nogtext=\"\" tdbgcolor=\"#f5f7ff\" thbgcolor=\"#000070\" thcolor=\"white\" checkcolor=\"black\" checksize=\"14\" arrowsize=\"14\" target=\"mail\" notopbuttons=\"\" columns=\"num,mark,new,answered,date,fromto,subject,size\" />\n" \
  "         </body>\n" \
  "        </else><comment>End IMHO main output</comment>\n" \
  "       </if><comment>End mail frame</comment>\n" \
--- 498,504 ----
  "         <if defined=\"imhoscreen is mailindex\">\n" \
  "          <h3><imho_mailbox />:</h3>\n" \
  "         </if>\n" \
! "         <imho_main nogtext=\"\" tdbgcolor=\"#3E60AA\" thbgcolor=\"#3E60AA\" thcolor=\"white\" checkcolor=\"black\" checksize=\"14\" arrowsize=\"14\" target=\"mail\" notopbuttons=\"\" columns=\"num,mark,new,answered,date,fromto,subject,size\" />\n" \
  "         </body>\n" \
  "        </else><comment>End IMHO main output</comment>\n" \
  "       </if><comment>End mail frame</comment>\n" \
***************
*** 502,508 ****
  "         <imho_name /><br /><hr />\n" \
  "         <table width=\"100%\" border=\"2\" cellspacing=\"0\" cellpadding=\"3\">\n" \
  "         <imho_buttonoutput target=\"mail\" quote=\"$\" noupdate=\"\">\n" \
! "           <tr><td bgcolor=#000070 align=\"center\">\n" \
  "           <a href=\"$button-url$\" target=\"$button-target$\" style=\"text-decoration: none\"><font color=white>$button-text$</font></a>\n" \
  "           </td></tr>\n" \
  "         </imho_buttonoutput>\n" \
--- 512,518 ----
  "         <imho_name /><br /><hr />\n" \
  "         <table width=\"100%\" border=\"2\" cellspacing=\"0\" cellpadding=\"3\">\n" \
  "         <imho_buttonoutput target=\"mail\" quote=\"$\" noupdate=\"\">\n" \
! "           <tr><td bgcolor=#3E60AAalign=\"center\">\n" \
  "           <a href=\"$button-url$\" target=\"$button-target$\" style=\"text-decoration: none\"><font color=white>$button-text$</font></a>\n" \
  "           </td></tr>\n" \
  "         </imho_buttonoutput>\n" \
***************
*** 515,521 ****
  "           </td></tr>\n" \
  "         </imho_buttonoutput>\n" \
  "	 <if supports=\"javascript\"><if defined=\"imhomailnotify is yes\">\n" \
! "          <tr><td bgcolor=\"#000070\" align=\"center\"><font color=\"white\">\n" \
  "           <cset preparse=\"\" scope=\"var\" variable=\"url\"><imho_url target=\"idlemailnotify\" href=\"checkactivemailboxes\" /></cset>\n" \
  "	   <script>\n" \
  "           function openWin(url,name,opts) {\n" \
--- 525,531 ----
  "           </td></tr>\n" \
  "         </imho_buttonoutput>\n" \
  "	 <if supports=\"javascript\"><if defined=\"imhomailnotify is yes\">\n" \
! "          <tr><td bgcolor=\"#3E60AA\" align=\"center\"><font color=\"white\">\n" \
  "           <cset preparse=\"\" scope=\"var\" variable=\"url\"><imho_url target=\"idlemailnotify\" href=\"checkactivemailboxes\" /></cset>\n" \
  "	   <script>\n" \
  "           function openWin(url,name,opts) {\n" \
***************
*** 533,539 ****
  "	   </font>\n" \
  "          </td></tr>\n" \
  "	 </if></if>\n" \
! "         <tr><td bgcolor=\"#000070\" align=\"center\">\n" \
  "         <imho_a href=\"reloadactivemailboxes\" target=\"mail\" style=\"text-decoration: none\"><font color=\"white\"><imho_string no=\"203\" /></font></a>\n" \
  "         </td></tr>\n" \
  "         </table>\n" \
--- 543,549 ----
  "	   </font>\n" \
  "          </td></tr>\n" \
  "	 </if></if>\n" \
! "         <tr><td bgcolor=\"#3E60AA\" align=\"center\">\n" \
  "         <imho_a href=\"reloadactivemailboxes\" target=\"mail\" style=\"text-decoration: none\"><font color=\"white\"><imho_string no=\"203\" /></font></a>\n" \
  "         </td></tr>\n" \
  "         </table>\n" \
***************
*** 545,555 ****
  " \n" \
  "         <if defined=\"imhoframe is list\">\n" \
  "          <h3><imho_mailbox />:</h3>\n" \
! "          <imho_mailindex nogtext=\"\" tdbgcolor=\"#f5f7ff\" thbgcolor=\"#000070\" thcolor=\"white\" checkcolor=\"black\" checksize=\"14\" arrowsize=\"14\" uptarget=\"mail\" sidetarget=\"main\" notopbuttons=\"\" columns=\"num,mark,new,answered,date,fromto,subject,size\" />\n" \
  "         </if><comment>End list frame</comment>\n" \
  "\n" \
  "         <else><comment>Finally! If no other frame, output imho_main</comment>\n" \
! "          <imho_main nogtext=\"\" tdbgcolor=\"#f5f7ff\" thbgcolor=\"#000070\" thcolor=\"white\" checkcolor=\"black\" checksize=\"14\" arrowsize=\"14\" target=\"mail\" nobackbutton=\"\" />\n" \
  "         </else>\n" \
  "        </else><comment>End list or main frames</comment>\n" \
  "        </body>\n" \
--- 555,565 ----
  " \n" \
  "         <if defined=\"imhoframe is list\">\n" \
  "          <h3><imho_mailbox />:</h3>\n" \
! "          <imho_mailindex nogtext=\"\" tdbgcolor=\"#3E60AA\" thbgcolor=\"#3E60AA\" thcolor=\"white\" checkcolor=\"black\" checksize=\"14\" arrowsize=\"14\" uptarget=\"mail\" sidetarget=\"main\" notopbuttons=\"\" columns=\"num,mark,new,answered,date,fromto,subject,size\" />\n" \
  "         </if><comment>End list frame</comment>\n" \
  "\n" \
  "         <else><comment>Finally! If no other frame, output imho_main</comment>\n" \
! "          <imho_main nogtext=\"\" tdbgcolor=\"#3E60AA\" thbgcolor=\"#3E60AA\" thcolor=\"white\" checkcolor=\"black\" checksize=\"14\" arrowsize=\"14\" target=\"mail\" nobackbutton=\"\" />\n" \
  "         </else>\n" \
  "        </else><comment>End list or main frames</comment>\n" \
  "        </body>\n" \
***************
*** 565,596 ****
  "   <body text=\"black\" bgcolor=\"#dde1f0\" link=\"black\" vlink=\"black\" alink=\"black\">\n" \
  "   <cset preparse=\"\" scope=\"var\" variable=\"banner\"><imho_banner /></cset>\n" \
  "   <if match=\"&roxen.version; is Roxen*\">\n" \
! "   <imho_image image=\"bannerleft\" text=\"&var.banner;\" quant=\"250\" clouds=\"white\" fg=\"white\" bg=\"#000070\" align=\"left\" border=\"0\" hspace=\"0\" />\n" \
  "   </if>\n" \
  "   <else>\n" \
  "   <formoutput quote=\"$\">\n" \
! "   <imho_image image=\"bannerleft\" text=\"$banner$\" quant=\"250\" clouds=\"white\" fg=\"white\" bg=\"#000070\" align=\"left\" border=\"0\" hspace=\"0\" /></formoutput>\n" \
  "   </else>\n" \
! "   <imho_image image=\"bannerright\" bg=\"#000070\" width=\"100%\" border=\"0\" hspace=\"0\" />\n" \
  "\n" \
  "   <table width=\"100%\" nowrap=\"nowrap\" cellpadding=\"2\" cellspacing=\"2\"><tr>\n" \
  "    <td bgcolor=\"#9099b4\">\n" \
  "     <table width=\"100%\"><tr><td>\n" \
  "      <table border=\"1\" cellpadding=\"5\" cellspacing=\"2\"><tr>\n" \
  "       <imho_buttonoutput quote=\"$\">\n" \
! "         <td bgcolor=#000070 align=\"center\">\n" \
  "         <a href=\"$button-url$\" target=\"$button-target$\" style=\"text-decoration: none\"><font color=white>$button-text$</font></a>\n" \
  "         </td>\n" \
  "       </imho_buttonoutput>\n" \
  "       </tr></table>\n" \
  "     </td></tr></table>\n" \
  "    </td>\n" \
! "    <td bgcolor=\"#000070\" align=\"center\">\n" \
  "      <font color=\"#ffffff\"><imho_name /> : <imho_title /></font>\n" \
  "    </td>\n" \
  "   </tr></table>\n" \
  "\n" \
! "   <imho_main nogtext=\"\" tdbgcolor=\"#f5f7ff\" thbgcolor=\"#000070\" thcolor=\"white\" checkcolor=\"black\" checksize=\"14\" arrowsize=\"14\" columns=\"num,mark,new,answered,date,fromto,subject,size\" />\n" \
  "   <hr /><imho_about />\n" \
  "   </body>\n" \
  "\n" \
--- 575,606 ----
  "   <body text=\"black\" bgcolor=\"#dde1f0\" link=\"black\" vlink=\"black\" alink=\"black\">\n" \
  "   <cset preparse=\"\" scope=\"var\" variable=\"banner\"><imho_banner /></cset>\n" \
  "   <if match=\"&roxen.version; is Roxen*\">\n" \
! "   <imho_image image=\"bannerleft\" text=\"&var.banner;\" quant=\"250\" clouds=\"white\" fg=\"white\" bg=\"#3E60AA\" align=\"left\" border=\"0\" hspace=\"0\" />\n" \
  "   </if>\n" \
  "   <else>\n" \
  "   <formoutput quote=\"$\">\n" \
! "   <imho_image image=\"bannerleft\" text=\"$banner$\" quant=\"250\" clouds=\"white\" fg=\"white\" bg=\"#3E60AA\" align=\"left\" border=\"0\" hspace=\"0\" /></formoutput>\n" \
  "   </else>\n" \
! "   <imho_image image=\"bannerright\" bg=\"#3E60AA\" width=\"100%\" border=\"0\" hspace=\"0\" />\n" \
  "\n" \
  "   <table width=\"100%\" nowrap=\"nowrap\" cellpadding=\"2\" cellspacing=\"2\"><tr>\n" \
  "    <td bgcolor=\"#9099b4\">\n" \
  "     <table width=\"100%\"><tr><td>\n" \
  "      <table border=\"1\" cellpadding=\"5\" cellspacing=\"2\"><tr>\n" \
  "       <imho_buttonoutput quote=\"$\">\n" \
! "         <td bgcolor=#3E60AAalign=\"center\">\n" \
  "         <a href=\"$button-url$\" target=\"$button-target$\" style=\"text-decoration: none\"><font color=white>$button-text$</font></a>\n" \
  "         </td>\n" \
  "       </imho_buttonoutput>\n" \
  "       </tr></table>\n" \
  "     </td></tr></table>\n" \
  "    </td>\n" \
! "    <td bgcolor=\"#3E60AA\" align=\"center\">\n" \
  "      <font color=\"#ffffff\"><imho_name /> : <imho_title /></font>\n" \
  "    </td>\n" \
  "   </tr></table>\n" \
  "\n" \
! "   <imho_main nogtext=\"\" tdbgcolor=\"#3E60AA\" thbgcolor=\"#3E60AA\" thcolor=\"white\" checkcolor=\"black\" checksize=\"14\" arrowsize=\"14\" columns=\"num,mark,new,answered,date,fromto,subject,size\" />\n" \
  "   <hr /><imho_about />\n" \
  "   </body>\n" \
  "\n" \
***************
*** 610,622 ****
  "  <body bgcolor=\"#dde1f0\" text=\"black\" link=\"black\" vlink=\"black\" alink=\"black\">\n" \
  "  <cset preparse=\"\" scope=\"var\" variable=\"banner\"><imho_banner /></cset>\n" \
  "  <if match=\"&roxen.version; is Roxen*\">\n" \
! "  <imho_image image=\"bannerleft\" text=\"&var.banner;\" quant=\"250\" clouds=\"white\" fg=\"white\" bg=\"#000070\" align=\"left\" border=\"0\" hspace=\"0\" /></formoutput>\n" \
! "  <imho_image image=\"bannerright\" bg=\"#000070\" width=\"100%\" border=\"0\" hspace=\"0\" />\n" \
  "  </if>\n" \
  "  <else>\n" \
  "  <formoutput quote=\"$\">\n" \
! "  <imho_image image=\"bannerleft\" text=\"$banner$\" quant=\"250\" clouds=\"white\" fg=\"white\" bg=\"#000070\" align=\"left\" border=\"0\" hspace=\"0\" /></formoutput>\n" \
! "  <imho_image image=\"bannerright\" bg=\"#000070\" width=\"100%\" border=\"0\" hspace=\"0\" />\n" \
  "  </else>\n" \
  "  <imho_main />\n" \
  "  <hr /><imho_about />\n" \
--- 620,632 ----
  "  <body bgcolor=\"#dde1f0\" text=\"black\" link=\"black\" vlink=\"black\" alink=\"black\">\n" \
  "  <cset preparse=\"\" scope=\"var\" variable=\"banner\"><imho_banner /></cset>\n" \
  "  <if match=\"&roxen.version; is Roxen*\">\n" \
! "  <imho_image image=\"bannerleft\" text=\"&var.banner;\" quant=\"250\" clouds=\"white\" fg=\"white\" bg=\"#3E60AA\" align=\"left\" border=\"0\" hspace=\"0\" /></formoutput>\n" \
! "  <imho_image image=\"bannerright\" bg=\"#3E60AA\" width=\"100%\" border=\"0\" hspace=\"0\" />\n" \
  "  </if>\n" \
  "  <else>\n" \
  "  <formoutput quote=\"$\">\n" \
! "  <imho_image image=\"bannerleft\" text=\"$banner$\" quant=\"250\" clouds=\"white\" fg=\"white\" bg=\"#3E60AA\" align=\"left\" border=\"0\" hspace=\"0\" /></formoutput>\n" \
! "  <imho_image image=\"bannerright\" bg=\"#3E60AA\" width=\"100%\" border=\"0\" hspace=\"0\" />\n" \
  "  </else>\n" \
  "  <imho_main />\n" \
  "  <hr /><imho_about />\n" \
***************
*** 637,650 ****
  " <body bgcolor=\"#dde1f0\" text=\"black\" link=\"black\" vlink=\"black\" alink=\"black\">\n" \
  " <cset preparse=\"\" scope=\"var\" variable=\"banner\"><imho_banner /></cset>\n" \
  " <if match=\"&roxen.version; is Roxen*\">\n" \
! " <imho_image image=\"bannerleft\" text=\"&var.banner;\" quant=\"250\" clouds=\"white\" fg=\"white\" bg=\"#000070\" align=\"left\" border=\"0\" hspace=\"0\" />\n" \
  " </if>\n" \
  " <else>\n" \
  " <formoutput quote=\"$\">\n" \
! " <imho_image image=\"bannerleft\" text=\"$banner$\" quant=\"250\" clouds=\"white\" fg=\"white\" bg=\"#000070\" align=\"left\" border=\"0\" hspace=\"0\" />\n" \
  " </formoutput>\n" \
  " </else>\n" \
! " <imho_image image=\"bannerright\" bg=\"#000070\" width=\"100%\" border=\"0\" hspace=\"0\" />\n" \
  " <center>\n" \
  "   <br /><h2><imho_banner /></h2>\n" \
  "   <imho_main nogtext=\"\" />\n" \
--- 647,660 ----
  " <body bgcolor=\"#dde1f0\" text=\"black\" link=\"black\" vlink=\"black\" alink=\"black\">\n" \
  " <cset preparse=\"\" scope=\"var\" variable=\"banner\"><imho_banner /></cset>\n" \
  " <if match=\"&roxen.version; is Roxen*\">\n" \
! " <imho_image image=\"bannerleft\" text=\"&var.banner;\" quant=\"250\" clouds=\"white\" fg=\"white\" bg=\"#3E60AA\" align=\"left\" border=\"0\" hspace=\"0\" />\n" \
  " </if>\n" \
  " <else>\n" \
  " <formoutput quote=\"$\">\n" \
! " <imho_image image=\"bannerleft\" text=\"$banner$\" quant=\"250\" clouds=\"white\" fg=\"white\" bg=\"#3E60AA\" align=\"left\" border=\"0\" hspace=\"0\" />\n" \
  " </formoutput>\n" \
  " </else>\n" \
! " <imho_image image=\"bannerright\" bg=\"#3E60AA\" width=\"100%\" border=\"0\" hspace=\"0\" />\n" \
  " <center>\n" \
  "   <br /><h2><imho_banner /></h2>\n" \
  "   <imho_main nogtext=\"\" />\n" \
***************
*** 660,665 ****
--- 670,677 ----

  // These properties are saved and loaded for each user. To add a new, change
  // here and in the SETUP menu.
+ // AJT 6Jul2000 sent-mail changed to sent
+ // AJT 25Sep2000 Trash changed to IMHO-Trash
  mapping prefproperties=([ "name" : "",
  			"address" : "",
  			"mailpath" : "",
***************
*** 671,678 ****
  			"visiblemail" : "20",
  			"sortorder" : "forward",
  			"sortcolumn" : "num",
! 			"trashfolder" : "Trash",
! 			"sentfolder" : "sent-mail",
  			"autobcc" : "",
  			"saveattachments" : "no",
  			"language" : "",
--- 683,690 ----
  			"visiblemail" : "20",
  			"sortorder" : "forward",
  			"sortcolumn" : "num",
! 			"trashfolder" : "IMHO-Trash",
! 			"sentfolder" : "sent",
  			"autobcc" : "",
  			"saveattachments" : "no",
  			"language" : "",
***************
*** 826,845 ****
  		     sprintf(" Sent mail from %s to %s, size %d\n",
  			     data->from, data->to, data->size));
      break;
    case "login":
      stats->nologins++;
      if (logfile && query("loglogin") == "yes")
!       logfile->write(now+" Login:"+data->login+"\n");
      break;
    case "loginfailed":
      stats->nofailedlogins++;
      if (logfile && query("loglogin") == "yes")
!       logfile->write(now+" Login failed:"+data->login+"\n");
      break;
    case "logout":
      stats->nologouts++;
      if (logfile && query("loglogin") == "yes")
!       logfile->write(now+" Logout:"+data->login+"\n");
      break;
    case "autologout":
      stats->noautologouts++;
--- 838,861 ----
  		     sprintf(" Sent mail from %s to %s, size %d\n",
  			     data->from, data->to, data->size));
      break;
+ //AJT 03Oct2000 added IP addr for logins (inc failed)
    case "login":
      stats->nologins++;
      if (logfile && query("loglogin") == "yes")
!       logfile->write(now+" Login:"+data->login+":"+data->ip+"\n");
! /*      logfile->write(now+" Login:"+data->login+"\n");*/
      break;
    case "loginfailed":
      stats->nofailedlogins++;
      if (logfile && query("loglogin") == "yes")
!       logfile->write(now+" Login failed:"+data->login+":"+data->ip+"\n");
! /*      logfile->write(now+" Login failed:"+data->login+"\n");*/
      break;
    case "logout":
      stats->nologouts++;
      if (logfile && query("loglogin") == "yes")
!       logfile->write(now+" Logout:"+data->login+":"+data->ip+"\n"); //AJT 10Oct2000
! /*      logfile->write(now+" Logout:"+data->login+"\n");*/
      break;
    case "autologout":
      stats->noautologouts++;
***************
*** 962,967 ****
--- 978,1003 ----
  #endif
  }

+ mapping mail_months = ([ "Jan":1, "Feb":2, "Mar":3, "Apr":4, "May":5, "Jun":6,
+ 			 "Jul":7, "Aug":8, "Sep":9, "Oct":10,"Nov":11,"Dec":12 ]);
+
+  string sortable_date(string s) {
+    int year,day,hour=-1,minute,second,timezone;
+    string wday,month;
+
+    if(!s) s="??????";
+
+    if (sscanf(s, "%s, %d %s %d %d:%d:%d",
+  	    wday, day, month, year, hour, minute, second)<4)
+      if (sscanf(s, "%d-%s-%d %d:%d:%d",
+  	       day, month, year, hour, minute, second, timezone)<6)
+        return s;
+
+    if (year<50) year+=2000;
+    if (year<1000) year+=1900;
+    return sprintf("%04d%02d%02d %02d%02d%02d", year, mail_months[month]||0, day, hour, minute, second);
+  }
+
  string parse_date(string s, void|int gtext){
    string ret="";
    int year,day,hour=-1,minute,second,timezone;
***************
*** 972,978 ****
  	    minute,second)<4)
      if(sscanf(s,"%d-%s-%d %d:%d:%d",day,month,year,hour,minute,second,timezone)<6)
        return s;
!   if (QUERY(shortdate))
      {
        year = year%100;
        ret+= sprintf("%02d",year);
--- 1008,1014 ----
  	    minute,second)<4)
      if(sscanf(s,"%d-%s-%d %d:%d:%d",day,month,year,hour,minute,second,timezone)<6)
        return s;
! /*  if (QUERY(shortdate))
      {
        year = year%100;
        ret+= sprintf("%02d",year);
***************
*** 981,989 ****
      {
        ret = sprintf("%02d",day);
        ret+= "-";
!     }

!   switch(month){
    case "Jan":
      ret+="01";
      break;
--- 1017,1025 ----
      {
        ret = sprintf("%02d",day);
        ret+= "-";
!     }*/

! /*  switch(month){
    case "Jan":
      ret+="01";
      break;
***************
*** 1023,1038 ****
    default:
      ret+="??";
      break;
!   }
!   if (!QUERY(shortdate))
      {
!       ret += sprintf("-%02d&nbsp;-&nbsp;%02d:%02d", year,hour,minute,second);
        // Work around for this dammed MSIE that have broken "non broken" space.
        if (!gtext)
  	{
  	  string bof="";
! 	  bof += "<if name=\"*MSIE*\">";
! 	  bof += "<gtext 5 nfont='Arial'>";
  	  bof += ret;
  	  bof += "</if>";
  	  bof += "<else>";
--- 1059,1116 ----
    default:
      ret+="??";
      break;
!   }*/
!    if (year<50) year+=2000;
!    if (year<1000) year+=1900;
!
!    switch(QUERY(dateformat)) {
!      case "ISO Short1":
!        ret=sprintf("%02d%02d%02d", year%100, mail_months[month]||0, day);
!        break;
!      case "ISO Short2":
!        ret=sprintf("%02d%02d%02d&nbsp;%02d%02d%02d",
!  		  year%100, mail_months[month]||0, day, hour, minute, second);
!        break;
!      case "ISO Long":
!        ret=sprintf("%04d-%02d-%02d&nbsp;%02d:%02d:%02d",
!  		  year, mail_months[month]||0, day, hour, minute, second);
!        break;
!      case "US Short1":
! /*       ret=sprintf("%d/%d/%d", day, month, year%100);*/
! /*       ret=sprintf("%d/%d/%02d", day, month, year%100);*/
! /*AJT minor fixes */
!        ret=sprintf("%d/%d/%02d", day, mail_months[month], year%100);
!        break;
!      case "US Short2":
! /*       ret=sprintf("%d/%d/%d&nbsp;%d:%02d:%02d&nbsp;%s",*/
! //AJT minor fixes
! //AJT 3Oct2000 12 midday should be 12 and not 0
!        int hour12=hour%12;
!        if(hour == 12)
! 		hour12=12;
!        ret=sprintf("%d/%d/%02d&nbsp;%d:%02d:%02d&nbsp;%s",
!  		  day, mail_months[month], year%100, hour12, minute, second,
!  		  (hour>12||hour==0)?"pm":"am");
!        break;
!      case "US Long":
!        ret=sprintf("%d&nbsp;%s&nbsp;%04d&nbsp;%d:%02d:%02d&nbsp;%s",
!  		  day, month, year, hour%12, minute, second,
!  		  (hour>12||hour==0)?"pm":"am");
!        break;
!      case "Old IMHO Long":
!        ret=sprintf("%02d-%02d-%04d&nbsp;-&nbsp;%02d:%02d:%02d",
!  		  day, month, year, hour, minute, second);
!        break;
!    }
! /*  if (!QUERY(shortdate))
      {
!       ret += sprintf("-%02d&nbsp;-&nbsp;%02d:%02d", year,hour,minute,second);*/
        // Work around for this dammed MSIE that have broken "non broken" space.
        if (!gtext)
  	{
  	  string bof="";
! 	  bof += "<if name=\"*aMSIE*\">";
! 	  bof += "<gtext 2 nfont='verdana , helvetica'>";
  	  bof += ret;
  	  bof += "</if>";
  	  bof += "<else>";
***************
*** 1040,1057 ****
  	  bof += "</else>";
  	  ret = bof;
  	}
!     }
    else
      {
        ret+=sprintf("%02d",day);
!     }
    return ret;
   }

  string make_links(object id, string text){
    string newstr="";
    //FIXME: 10000 ?
!   string instr=text[0..(10000<sizeof(text)?10000:sizeof(text)-1)];

    array (string)keys=({ "http://" , "ftp://" , "telnet://","gopher://","wais://", "@","https://" });
    string stoptkn=" \n\r\"()[]{}<>'`;";
--- 1118,1137 ----
  	  bof += "</else>";
  	  ret = bof;
  	}
! /*    }
    else
      {
        ret+=sprintf("%02d",day);
!     }*/
    return ret;
   }

  string make_links(object id, string text){
    string newstr="";
    //FIXME: 10000 ?
!   int big_mail= 50000; // AJT 15Sep2000
! /*  string instr=text[0..(10000<sizeof(text)?10000:sizeof(text)-1)];*/
!   string instr=text[0..(big_mail<sizeof(text)?big_mail:sizeof(text)-1)];

    array (string)keys=({ "http://" , "ftp://" , "telnet://","gopher://","wais://", "@","https://" });
    string stoptkn=" \n\r\"()[]{}<>'`;";
***************
*** 1112,1119 ****
    }
    newstr+=instr[position..];
    //FIXME: 10000 ?
!   if (sizeof(text) > 10001)
!     newstr += text[10001..sizeof(text)-1];
    return newstr;
  }

--- 1192,1201 ----
    }
    newstr+=instr[position..];
    //FIXME: 10000 ?
! /*  if (sizeof(text) > 10001)*/
!   if (sizeof(text) > big_mail+1)
!     newstr += text[(big_mail+1)..sizeof(text)-1];
! /*    newstr += text[10001..sizeof(text)-1];*/
    return newstr;
  }

***************
*** 1149,1154 ****
--- 1231,1238 ----

  // Make sure the all email addresses are on correct form
  array fixaddresses(string addr) {
+   string rfc822_specials ="()<>@,.;:[]\\\""; //AJT 19Aug2000
+   int rfc822_quote=0; //AJT
    array ret = ({ "", "" });
    foreach(address_split(addr), string s) {
      string fixedaddr = "";
***************
*** 1201,1207 ****
        fixedaddr = address;
      else {
        name = replace(name, "\"", "");
!       fixedaddr = "\""+name+"\" <"+address+">";
      }
      if (sizeof(ret[0]) == 0)
        ret[0] = fixedaddr;
--- 1285,1300 ----
        fixedaddr = address;
      else {
        name = replace(name, "\"", "");
! // AJT 19Aug2000 rfc822 specials only quoted
!       for(int jj=0;jj<sizeof(name);jj++) {
! 	string jj2=name[jj..jj];
! 	if(search(rfc822_specials,jj2)>= 0)
! 		rfc822_quote=1;
! 		}
!       if(rfc822_quote)
! 	      fixedaddr = "\""+name+"\" <"+address+">";
!       else
! 	      fixedaddr = ""+name+" <"+address+">";
      }
      if (sizeof(ret[0]) == 0)
        ret[0] = fixedaddr;
***************
*** 1755,1760 ****
--- 1848,1856 ----
      {
        if (!s)
  	return "";
+ //AJT hack 7Sep2000 - leading newline appears and must be removed
+      if(s[0..0] == "\n")
+ 	s = s[1..];
        return replace(s, ({ "<br", "\n", "\r" }), ({ "<br0", "<br>", "" }));
      }

***************
*** 1824,1830 ****
      }
      if(sessobj->status<0) {
        if (command_result == -1) {
! 	GLOBAL->imho_log("loginfailed", ([ "login":sessobj->address ]));
  	sessobj->status=LOGINFAILED;
        }
        else
--- 1920,1928 ----
      }
      if(sessobj->status<0) {
        if (command_result == -1) {
! /*	GLOBAL->imho_log("loginfailed", ([ "login":sessobj->address ]));*/
! //AJT 03Oct2000 log IP as well
! 	GLOBAL->imho_log("loginfailed", ([ "login":sessobj->address,"ip":sessobj->ip ]));
  	sessobj->status=LOGINFAILED;
        }
        else
***************
*** 3462,3467 ****
--- 3560,3573 ----
  	      }
  	    }
  	    break;
+ #if constant(Protocols.LDAP)
+           case "ldap":
+             name=getfullnamefromldap(sessobj->defaultaddress);
+             if(sizeof(name) > 1) { // Only set name if something was returned
+               sessobj->name=name;
+             }
+             break;
+ #endif
  	  default:
  	    break;
  	  }
***************
*** 3472,3478 ****
        break;

      case "start":
!       GLOBAL->imho_log("login", ([ "login":sessobj->address ]));
        if(sessobj->prefsloaded || !sessobj->usersetup)
  	(sessobj)["status"] = MAILINDEX;
        else
--- 3578,3586 ----
        break;

      case "start":
! /*      GLOBAL->imho_log("login", ([ "login":sessobj->address ]));*/
! //AJT 03Oct2000 log IP as well
!       GLOBAL->imho_log("login", ([ "login":sessobj->address,"ip":sessobj->ip ]));
        if(sessobj->prefsloaded || !sessobj->usersetup)
  	(sessobj)["status"] = MAILINDEX;
        else
***************
*** 3737,3743 ****
      return "";
    return sprintf("%s, %02d %s %d %02d:%02d:%02d %s%04d",
  		 wday, lt->mday, month, lt->year+1900, lt->hour, lt->min,
! 		 lt->sec, -lt->timezone >=0 ? "+":"", -lt->timezone*100/3600);
  }

  string
--- 3845,3854 ----
      return "";
    return sprintf("%s, %02d %s %d %02d:%02d:%02d %s%04d",
  		 wday, lt->mday, month, lt->year+1900, lt->hour, lt->min,
! /*		 lt->sec, -lt->timezone >=0 ? "+":"", -lt->timezone*100/3600);
! -                lt->sec, -lt->timezone >=0 ? "+":"", -lt->timezone*100/3600);*/
!                 lt->sec, -lt->timezone >=0 ? "+":"",
!                 (lt->isdst==1?100:0)-lt->timezone*100/3600);
  }

  string
***************
*** 3864,3869 ****
--- 3975,3991 ----
    //FIXME: doesn't honour the length limits, 75,76 chars.
    //       it would be very difficult due to the integral constraint
    //       for multibyte character sets
+ //look for high bit chars in s1 14Sep2000 AJT
+ //originally in encode_addresses moved here 22Oct2000
+ //if not found just return address ASIS
+ int high_bit=0;
+     for(int jj=0;jj<sizeof(s);jj++) {
+         if(s[jj] > 127)
+                 high_bit=1;
+         }
+   if(high_bit) {
+ //replace comma with space since encode fails to convert comma 14Sep2000
+       s = replace(s,","," ");
    array a1=(Locale.Charset.encoder(cs,"")->feed(s)->drain())/" ";
    array a2=Array.map(a1,lambda(string s,string cs){return MIME.encode_word(({s,cs}),"quoted-printable");},cs);
    array a3=Array.map(a2,lambda(string s){return (s/"?");} );
***************
*** 3879,3884 ****
--- 4001,4009 ----
      }
    }
    return a1*" ";
+   }
+ else
+ 	return s;
  }


***************
*** 4590,4596 ****
  #endif
  	 "User names:Method", TYPE_STRING_LIST,
  	 "Chooses method to set the users names.",
!          ({ "file", "auth module","none" }));

    defvar("namedb",
  #ifndef __NT__
--- 4715,4725 ----
  #endif
  	 "User names:Method", TYPE_STRING_LIST,
  	 "Chooses method to set the users names.",
!          ({ "file", "auth module","none"
! #if constant(Protocols.LDAP)
!             ,"ldap"
! #endif
!             }));

    defvar("namedb",
  #ifndef __NT__
***************
*** 4791,4799 ****
  	 "mail being sent as UTF-8.");


!   defvar("shortdate",1,"User interface dialogs:Folder index - Use short date", TYPE_FLAG,
  	 "When set, the date is presented as YYMMDD, otherwise the full date is shown (european format) as : "
! 	 "DD:MM:YYYY - HH:MM:SS instead");


    defvar("signaturecols","55","User interface dialogs: Pages - Setup page, size of signature", TYPE_STRING,
--- 4920,4938 ----
  	 "mail being sent as UTF-8.");


! /*  defvar("shortdate",1,"User interface dialogs:Folder index - Use short date", TYPE_FLAG,
  	 "When set, the date is presented as YYMMDD, otherwise the full date is shown (european format) as : "
! 	 "DD:MM:YYYY - HH:MM:SS instead");*/
!    defvar("dateformat", "ISO Short1" ,"User interface dialogs:Folder index - date format", TYPE_STRING_LIST,
!  	 "The desired date format in the Folder index:<br />\n"
!  	 "<b>ISO Short1</b>&nbsp;&nbsp;&nbsp;YYMMDD (e.g 981221)<br />\n"
!  	 "<b>ISO Short2</b>&nbsp;&nbsp;&nbsp;YYMMDD HHMMSS (e.g 981221 18:19:20)<br />\n"
!  	 "<b>ISO Long</b>&nbsp;&nbsp;&nbsp;YYYY-MM-DD HH:MM:SS (eg. 2000-01-17 17:47:11)<br />\n"
!  	 "<b>US Short1</b>&nbsp;&nbsp;&nbsp;D/M/Y (eg. 9/3/99)<br />\n"
!  	 "<b>US Short2</b>&nbsp;&nbsp;&nbsp;D/M/Y H:MM:SS XX (eg. 19/03/99 5:12:54 pm)<br />\n"
!  	 "<b>US Long</b>&nbsp;&nbsp;&nbsp;D Mon YYYYY H:MM:SS XX (eg. 4 Jul 2001 11:01:01 am)<br />\n"
!  	 "<b>Old IMHO Long</b>&nbsp;&nbsp;&nbsp;DD-MM-YYYY - HH:MM:SS (eg 15-07-1999 17:42:01)<br />\n",
!  	 ({ "ISO Short1","ISO Short2", "ISO Long", "US Short1","US Short2", "US Long", "Old IMHO Long" }));


    defvar("signaturecols","55","User interface dialogs: Pages - Setup page, size of signature", TYPE_STRING,
***************
*** 4947,4960 ****
    return(ab);
  }

  string import_addressbook(string file) {
    string ab = "";
    string adds = replace(file, "\r", "");
    array (string) lines = adds / "\n";
    if (sizeof(lines) > 0) {
      string name="", address="";
!     if (sscanf(lines[0], "%s\t%s", name, address) == 2)
        ab = import_pinebook(adds);
      else if (sscanf(lines[0], "dn: %s", address) == 1)
        ab = import_ldifbook(adds);
    }
--- 5086,5139 ----
    return(ab);
  }

+ string pre_pinebook(string adds) {
+ //function to pre-process pine 4.x address books older format...
+ //AJT 3Sep2000
+ array (string) l = adds / "\n";
+ array (string) lines = allocate(sizeof(l)); // create array large enough
+ //show line by line
+ int i=0;
+ string oneline;
+ foreach(l,oneline) {
+ //weed out comments
+ 	if(oneline[0..0] == "#") {
+ 		continue;
+ 		}
+ 	if(oneline[0..0] == " " ) {
+ 		//lines[i-1] = lines[i-1]+oneline;
+ 		lines[i-1] += oneline;
+ 		}
+ 	else {
+ 		lines[i]=oneline;
+ 		i++;
+ 		}
+ 	}
+ adds="";
+ for(int j=0;j<i-1;j++) {
+ 	array (string) addlins = lines[j]/"\t";
+ 	if(String.trim_whites(addlins[1])=="") {
+ 		addlins[1] = addlins[0];
+ 		}
+ 	//add newlines...
+ 	adds+=addlins[0]+"\t"+addlins[1]+"\t"+String.trim_whites(addlins[2])+"\n";
+ 	}
+ 	return(adds);
+ }
  string import_addressbook(string file) {
    string ab = "";
    string adds = replace(file, "\r", "");
    array (string) lines = adds / "\n";
    if (sizeof(lines) > 0) {
      string name="", address="";
! /*    if (sscanf(lines[0], "%s\t%s", name, address) == 2)*/
! //changed AJT 4,5Sep2000
!       if (sscanf(lines[0], "Name=%s", name) == 1) //must come before pine
! 	      ab = import_simeon_book(adds); // read Simeon address book
!     else if (sscanf(lines[0], "%s\t%s", name, address) == 2) {
! // pre-process data AJT 3Sep2000 for pine 4.x address books
! 	adds = pre_pinebook(adds); //added function
        ab = import_pinebook(adds);
+ 	}
      else if (sscanf(lines[0], "dn: %s", address) == 1)
        ab = import_ldifbook(adds);
    }
***************
*** 5010,5015 ****
--- 5189,5243 ----
    return(ab);
  }

+ string import_simeon_book(string adds) {
+ //read exported simeon address book
+ //AJT 5Sep2000,12Sep2000
+ string ab="";
+ 	mapping(string:string) map=([]);
+ //split into lines
+ 	array (string) lines = adds/"\n";
+ 	string name="";
+ 	string address="",list="";
+ 	for(int j=0;j<sizeof(lines)-1;j++) {
+ 	array (string) comps = lines[j]/"\t";
+ 	name=String.trim_whites(comps[0][5..]);
+ //split into single addresses and lists
+ // look for email tag or member tag, if neither is found junk line 12Sep2000
+ 	int address_set=0;
+ 	for(int p=0;p<sizeof(comps);p++) {
+ 	if(lower_case(comps[p][0..4]) == "email") {
+ 		address=comps[p][6..];
+ 		address_set = 1;
+ 		}
+ 	if(lower_case(comps[p][0..5]) == "member" ) {
+ 		address=comps[p][7..];
+ 		address_set=1;
+ 		}
+ 	}
+ 	if(address_set == 0 )
+ 		continue; // next loop iteration
+ 	map[name]=address;
+ 	}
+ //2nd loop to fully expand lists...
+ 	foreach(indices(map),string person) {
+ 		if(search(map[person],",") != -1) {
+ 			array(string) listcomps = map[person]/",";
+ 			string list="";
+ 			for(int k=0;k<sizeof(listcomps);k++) {
+ //is address already a LHS value, if so expand
+ 				if(map[listcomps[k]] != 0)
+ 					listcomps[k]=map[listcomps[k]];
+ 				}
+ 			list=listcomps*",";
+ 			ab += person+":"+list+"\n";
+ 			}
+ 		else {
+ 			ab += person+":"+map[person]+"\n";
+ 			}
+ 		}
+ 	ab=ab[0..sizeof(ab)-2]; //remove last LF
+ 	return(ab);
+ }

  string status() {
    return "Sessions: "+sizeof(sessions)+"<br />"
***************
*** 5367,5373 ****
        return id->misc->imho->sessobj->old_module->tag_imho_main(tag_name,arg,id,file);

      string thcolor=arg->thcolor || "#ffffff";
!     string thbgcolor=arg->thbgcolor || "#000070";
      string tdbgcolor=arg->tdbgcolor || "#fcfce0";
      string checkcolor=arg->checkcolor || "#000000";
      int checksize=(int)arg->checksize || 15;
--- 5595,5601 ----
        return id->misc->imho->sessobj->old_module->tag_imho_main(tag_name,arg,id,file);

      string thcolor=arg->thcolor || "#ffffff";
!     string thbgcolor=arg->thbgcolor || "#3E60AA";
      string tdbgcolor=arg->tdbgcolor || "#fcfce0";
      string checkcolor=arg->checkcolor || "#000000";
      int checksize=(int)arg->checksize || 15;
***************
*** 5889,5894 ****
--- 6117,6171 ----
    return FROM_UTF8(output);
  }

+ // Queries an LDAP server for the users fullname, based on a username.
+ string getfullnamefromldap(string login)
+ {
+   string filter, server, basedn, name;
+   mixed status;
+   mixed error;
+   object con;
+   object en;
+
+   server = query("ldapserver");
+   if (server == "")
+     server = "localhost";
+   basedn = query("ldapsearchroot");
+
+   error = catch(con = Protocols.LDAP.client(server));
+   error |= catch(con->bind());
+
+   if (error || !objectp(con)) {
+     if (query("debug")=="on")
+       perror("couldn't connect to LDAP-server "+server+"\n");
+     return "";
+   }
+
+   status = con->set_basedn(basedn);
+   status = con->set_scope(2);
+
+   filter = "(mail="+login+")";
+   if (error = catch(en = con->search(filter))) {
+     if (query("debug")=="on")
+       perror("LDAP search failed "+con->error_string()+"\n");
+     return "";
+   }
+   con->unbind();
+
+   // return null if not exactly match 1 entry
+   if (en->num_entries() == 1) {
+       name = (en->fetch(1)["cn"]||({ "" }))[0];
+       if(sizeof(name) == 0) {
+         name = (en->fetch(1)["givenname"]||({ "" }))[0] + " " + (en->fetch(1)["sn"]||({ "" }))[0];
+       }
+       return FROM_UTF8(name);
+   }
+   else {
+     if (query("debug")=="on")
+       perror("none or more than 1 entry returned\n");
+     return "";
+   }
+ }
+
  // Queries and LDAP server of which IMAP server to use, based on a username
  string getimapserver(string login)
  {
***************
*** 5975,5981 ****
  #endif
      ret+"<br /><br />";
      ret+=MSG(M_LOGOUTMSG);
!     ret+="<remove_cookie name=IMHOSession>";
      ret+="</body></html>";
      return ret;
      break;
--- 6252,6258 ----
  #endif
      ret+"<br /><br />";
      ret+=MSG(M_LOGOUTMSG);
!     ret+="<remove-cookie name=IMHOSession>";
      ret+="</body></html>";
      return ret;
      break;
***************
*** 6016,6022 ****
    ret+="<table width=100%><tr>";
    array columns = ({ "Login", "Login Time", "Idle Time", "IP-Address", "# Msgs in Box", "Layout", "Language" });
    for (int i = 0; i < sizeof(columns); i++) {
!     ret+="<th bgcolor=#000070><font color=white><b><a href=\""+id->misc->imho->nexturl+"?actionchangesort=1&col="+i+"&dir="+(sessobj->sortup==0&&sessobj->adminsortcolumn==i?"1":"0")+"\"><imho_image bg=\"#000070\" border=\"0\" alt=\""+MSG(M_CHANGESORTORDER)+"\" size=\"60\" scale=\"0.25\" image=\""+((sessobj->adminsortcolumn==i)?(sessobj->sortup?"arrowup":"arrowdown"):"arrownone")+"\" /></a>&nbsp;"+columns[i]+"</b></font></th>";
    }
    ret+="</tr>\n";

--- 6293,6299 ----
    ret+="<table width=100%><tr>";
    array columns = ({ "Login", "Login Time", "Idle Time", "IP-Address", "# Msgs in Box", "Layout", "Language" });
    for (int i = 0; i < sizeof(columns); i++) {
!     ret+="<th bgcolor=#3E60AA><font color=white><b><a href=\""+id->misc->imho->nexturl+"?actionchangesort=1&col="+i+"&dir="+(sessobj->sortup==0&&sessobj->adminsortcolumn==i?"1":"0")+"\"><imho_image bg=\"#3E60AA\" border=\"0\" alt=\""+MSG(M_CHANGESORTORDER)+"\" size=\"60\" scale=\"0.25\" image=\""+((sessobj->adminsortcolumn==i)?(sessobj->sortup?"arrowup":"arrowdown"):"arrownone")+"\" /></a>&nbsp;"+columns[i]+"</b></font></th>";
    }
    ret+="</tr>\n";

***************
*** 6186,6191 ****
--- 6463,6474 ----

      case "login":
        if(sessobj->status < 0) {
+ //AJT
+ 	if(!sessobj->cookies) {
+ 		(sessobj)["status"] = LOGIN_NOCOOKIE;
+ 		break;
+ 		}
+ //AJT end
  	sessobj->logintime = time();
  	array loginsplit = id->variables->login / "@";
  	sessobj["login"] = loginsplit[0];
***************
*** 6296,6302 ****

      case "logout":
        (sessobj)["status"] = LOGOUT;
!       imho_log("logout", ([ "login":sessobj->address ]));
        sessobj->imapclient->imap_command(id,sessobj,({ imap_cmd("low_logout", "noselect", 1) }));
        return http_pipe_in_progress();
        break;
--- 6579,6587 ----

      case "logout":
        (sessobj)["status"] = LOGOUT;
! //AJT 10Oct2000 show IP on logout
!       imho_log("logout", ([ "login":sessobj->address,"ip":sessobj->ip ]));
! /*      imho_log("logout", ([ "login":sessobj->address ]));*/
        sessobj->imapclient->imap_command(id,sessobj,({ imap_cmd("low_logout", "noselect", 1) }));
        return http_pipe_in_progress();
        break;
***************
*** 6943,6955 ****
--- 7228,7253 ----
        if(id->variables->mbox2 && id->variables->mbox2!="imhonomailbox")
  	mboxtomoveto=id->variables->mbox2;
      case "move":
+ // code added AJT 19Sep2000 from delete checks to avoid bug if user
+ // clicks on move without marking
+       if (sizeof(sessobj->selected) > 0) { //have we marked anything?
        if(!mboxtomoveto && id->variables->mbox && id->variables->mbox!="imhonomailbox")
  	mboxtomoveto=id->variables->mbox;
        if(mboxtomoveto) {
  	sessobj["copytobox"]=mboxtomoveto;
  	sessobj->imapclient->imap_command(id,sessobj,({ imap_cmd("move","uids",sessobj->selected+({ }),"updatembox",imap_cmd_var("mails") ) }));
  	return http_pipe_in_progress();
+        }
        }
+       else {
+       sessobj["dialogstrings"] = ({ MSG(M_DIALOGOK) });
+       sessobj["dialogactions"] = ({ "actionindex" });
+       sessobj["dialogtarget"] = id->misc->imho->frame;
+       sessobj["dialogtext"] =
+ 	MSG(M_DELETEMARKEDNONE);
+       (sessobj)["status"] = DIALOGBOX;
+ 	}
+       break;
        break;

      case "reload":
***************
*** 6991,6997 ****
  	sessobj->copytobox=encode_mboxname(sessobj->copytobox);
        int prevuid=(int)id->variables->prevuid || -1;
        int nextuid=(int)id->variables->nextuid || -1;
!       int uid=(nextuid==-1?prevuid:nextuid);
        sessobj->cmailidx = find_mail(sessobj->mails, "UID", uid);
        sessobj->cmailuid = uid;
        if (uid ==-1) {
--- 7289,7301 ----
  	sessobj->copytobox=encode_mboxname(sessobj->copytobox);
        int prevuid=(int)id->variables->prevuid || -1;
        int nextuid=(int)id->variables->nextuid || -1;
! //AJT 04Oct2000 next uid dependant or sortorder
! /*      int uid=(nextuid==-1?prevuid:nextuid);*/
! int uid; //important to do this here
! if(sessobj->sessionsortorder == "forward")
!       uid=(nextuid==-1?prevuid:nextuid); //initial forward default
! else
!       uid=(nextuid==-1?nextuid:prevuid);
        sessobj->cmailidx = find_mail(sessobj->mails, "UID", uid);
        sessobj->cmailuid = uid;
        if (uid ==-1) {
***************
*** 7006,7012 ****
      case "deletethis":
        int prevuid=(int)id->variables->prevuid || -1;
        int nextuid=(int)id->variables->nextuid || -1;
!       int uid=(nextuid==-1?prevuid:nextuid);
        sessobj->cmailidx = find_mail(sessobj->mails, "UID", uid);
        sessobj->cmailuid = uid;
        if (uid == -1) {
--- 7310,7316 ----
      case "deletethis":
        int prevuid=(int)id->variables->prevuid || -1;
        int nextuid=(int)id->variables->nextuid || -1;
!       int uid=(nextuid==-1?prevuid:nextuid);
        sessobj->cmailidx = find_mail(sessobj->mails, "UID", uid);
        sessobj->cmailuid = uid;
        if (uid == -1) {
***************
*** 7130,7143 ****
        (sessobj)["subject"] = fix_coding(mail->headers->subject);
        if ((sessobj)["subject"][0..3] != "Fwd:")
  	(sessobj)["subject"] = "Fwd: "+(sessobj)["subject"];
        (sessobj)["message"]+="\n\nFrom: "+fix_coding(mail->headers->from)+"\n";
        if(mail->headers["reply-to"])
! 	(sessobj)["message"]+="Reply-to: "+ fix_coding(mail->headers["reply-to"]);
        (sessobj)["message"]+="To: "+fix_coding(mail->headers->to)+"\n";
        if(mail->headers->cc)
  	(sessobj)["message"]+="Cc: "+ fix_coding(mail->headers->cc )+"\n";
        (sessobj)["message"]+="Date: "+ fix_coding(mail->headers->date || "");
        (sessobj)["message"]+="\n\n";


        if(mail->body_parts){
--- 7434,7464 ----
        (sessobj)["subject"] = fix_coding(mail->headers->subject);
        if ((sessobj)["subject"][0..3] != "Fwd:")
  	(sessobj)["subject"] = "Fwd: "+(sessobj)["subject"];
+ // AJT 16Aug2000 add all hdrs if show all headers is set...
+ // based on stewa@lysator.liu.se Sun Jul  2 20:41:16 2000 to get all hdrs
+       if(!sessobj->showheaders) {
        (sessobj)["message"]+="\n\nFrom: "+fix_coding(mail->headers->from)+"\n";
        if(mail->headers["reply-to"])
! 	(sessobj)["message"]+="Reply-to: "+ fix_coding(mail->headers["reply-to"])+"\n"; //minor fix by AJT 17Aug2000
! /*	(sessobj)["message"]+="Reply-to: "+ fix_coding(mail->headers["reply-to"]);*/
        (sessobj)["message"]+="To: "+fix_coding(mail->headers->to)+"\n";
        if(mail->headers->cc)
  	(sessobj)["message"]+="Cc: "+ fix_coding(mail->headers->cc )+"\n";
        (sessobj)["message"]+="Date: "+ fix_coding(mail->headers->date || "");
        (sessobj)["message"]+="\n\n";
+ 	}
+       else {
+ // put all headers in new mail, this a hack of the full hdr dipslay code
+       foreach(indices(mail->headers), string h) {
+         if (!(<"">)[h])
+           foreach(mail->headers[h]/"\0",string h2)
+             (sessobj)["message"]+= String.capitalize(h)+": "+h2+"\n";
+ /*            MAIN+="<br /><b>"+h+":</b> "+fix_coding(h2);*/
+ 		}
+ // add extra newlines
+ 	(sessobj)["message"] += "\n";
+ 	}
+ // end of new stuff


        if(mail->body_parts){
***************
*** 7486,7497 ****
--- 7807,7824 ----
    case LOGINFAILEDIMAP:
    case LOGINFAILED:
    case LOGINPROMPT:
+ //AJT
+   case LOGIN_NOCOOKIE:
      id->misc->imho->screen="login";
      if(screen==LOGINFAILED)
        MAIN+=MSG(M_NOLOGIN)+"<br />\n<br />\n";
      else
        if(screen==LOGINFAILEDIMAP)
  	MAIN+=MSG(M_IMAPERROR)+"<br />\n<br />\n";
+ //AJT
+     else
+ 	if(screen == LOGIN_NOCOOKIE)
+ 	MAIN+=MSG(M_NOCOOKIE)+"<br />\n<br />\n";
      string preset_login=0;
      if(plugin && plugin->imho_preset_login)
        preset_login=plugin->imho_preset_login(id);
***************
*** 7531,7537 ****
  	MAIN+="document.imhologinform.login.focus();\n";
        MAIN+="// -->\n</script>\n";
      }
!     MAIN+="<set_cookie name=IMHOSession value="+sessobj->session+">";
      break;


--- 7858,7864 ----
  	MAIN+="document.imhologinform.login.focus();\n";
        MAIN+="// -->\n</script>\n";
      }
!     MAIN+="<set-cookie name=IMHOSession value="+sessobj->session+">";
      break;


***************
*** 7540,7546 ****
      object lock;
      id->misc->imho->title=MSG(M_LOGGEDOUT);
      MAIN+=MSG(M_LOGOUTMSG);
!     MAIN+="<remove_cookie name=IMHOSession>";
  #if THREADS
      lock = global_lock->lock();
  #endif
--- 7867,7873 ----
      object lock;
      id->misc->imho->title=MSG(M_LOGGEDOUT);
      MAIN+=MSG(M_LOGOUTMSG);
!     MAIN+="<remove-cookie name=IMHOSession>";
  #if THREADS
      lock = global_lock->lock();
  #endif
***************
*** 7756,7763 ****
        MAIN+="<th class=\"ou\" bgcolor=\"&imho_thbgcolor;\"><font color=\"&imho_thcolor;\">"+MSG(M_OU)+"</font></th>";
      MAIN+="<tr>";
      int i = 0;
!     if (sizeof(sessobj->ldapadd)==0)
        MAIN+="<tr><td>"+MSG(M_NOADDRESSES)+"</td></tr>";
      foreach(sessobj->ldapadd/"\n", string line) {
        string name = "", address = "", ou="";
        if(query("ldapshowou")=="yes") {
--- 8083,8096 ----
        MAIN+="<th class=\"ou\" bgcolor=\"&imho_thbgcolor;\"><font color=\"&imho_thcolor;\">"+MSG(M_OU)+"</font></th>";
      MAIN+="<tr>";
      int i = 0;
! /*    if (sizeof(sessobj->ldapadd)==0)*/
!     if (sizeof(sessobj->ldapadd)==0) {
        MAIN+="<tr><td>"+MSG(M_NOADDRESSES)+"</td></tr>";
+ //AJT back to compose link 24Sep2000
+       MAIN+="<tr><td><a target=\""+id->misc->imho->target+"\" href=\""+id->misc->imho->nexttarget;
+       MAIN+="?actioncontinuecompose=1\"/a>Back to Compose</td></tr>";
+ //end add
+ 	}
      foreach(sessobj->ldapadd/"\n", string line) {
        string name = "", address = "", ou="";
        if(query("ldapshowou")=="yes") {
***************
*** 7854,7860 ****
  	  object mail = (sessobj)["mails"][i];
   	  switch(sessobj->sortcolumn) {
   	  case "date":
!  	    colvals[parse_date(mail->imap->INTERNALDATE)+" "+(string) sprintf("%08d", mail->number)]=i;
   	    break;
   	  case "from":
   	    string from_name,from_addr;
--- 8187,8194 ----
  	  object mail = (sessobj)["mails"][i];
   	  switch(sessobj->sortcolumn) {
   	  case "date":
! /* 	    colvals[parse_date(mail->imap->INTERNALDATE)+" "+(string) sprintf("%08d", mail->number)]=i;*/
!   	    colvals[sortable_date(mail->imap->INTERNALDATE)+" "+(string) sprintf("%08d", mail->number)]=i;
   	    break;
   	  case "from":
   	    string from_name,from_addr;
***************
*** 7965,7973 ****
--- 8299,8313 ----

  	mailnumbers+=mnrs+" ";
  	if (sessobj->firstvisiblemail > 0)
+       if (sessobj->sessionsortorder == "forward")
  	  mailnumbers += " <a href=\""+id->misc->imho->nextpage+"?actionprevblock=1\">"+html_encode_string(MSGA(M_BACKN, ({ sessobj->visible })))+"</a>";
+       else
+ 	  mailnumbers += " <a href=\""+id->misc->imho->nextpage+"?actionprevblock=1\">"+html_encode_string(MSGA(M_FORWARDN, ({ sessobj->visible })))+"</a>";
  	if (sessobj->lastvisiblemail+1 < mails)
+      if (sessobj->sessionsortorder == "forward")
  	  mailnumbers += " <a href=\""+id->misc->imho->nextpage+"?actionnextblock=1\">"+html_encode_string(MSGA(M_FORWARDN, ({ sessobj->visible })))+"</a>";
+      else
+ 	  mailnumbers += " <a href=\""+id->misc->imho->nextpage+"?actionnextblock=1\">"+html_encode_string(MSGA(M_BACKN, ({ sessobj->visible })))+"</a>";
  	mailnumbers += "</td></tr>\n</table>";
        }
        if (!id->misc->imho->notopbuttons)
***************
*** 8022,8028 ****
  	  MAIN+="<th class=\"answered\" bgcolor=\"&imho_thbgcolor;\" valign=\"middle\"><font class=\"answered\" color=\"&imho_thcolor;\">"+MSG(M_ANSWEREDFLAG)+"</font></th>";
  	  break;
  	case "date":
! 	  MAIN+="<th class=\"date\" bgcolor=\"&imho_thbgcolor;\" valign=\"middle\">";
  	  if (sessobj->sortcolumn == column)
  	    MAIN+=sortarrow;
  	  else
--- 8362,8368 ----
  	  MAIN+="<th class=\"answered\" bgcolor=\"&imho_thbgcolor;\" valign=\"middle\"><font class=\"answered\" color=\"&imho_thcolor;\">"+MSG(M_ANSWEREDFLAG)+"</font></th>";
  	  break;
  	case "date":
! 	  MAIN+="<th class=\"date\" bgcolor=\"&imho_thbgcolor;\" width=\"20%\"  valign=\"middle\">";
  	  if (sessobj->sortcolumn == column)
  	    MAIN+=sortarrow;
  	  else
***************
*** 8061,8067 ****
  	  MAIN+="<font class=\"to\" color=\"&imho_thcolor;\">"+MSG(M_TO)+"</font></th>";
  	  break;
  	case "subject":
! 	  MAIN+="<th class=\"subject\" bgcolor=\"&imho_thbgcolor;\" width=\"60%\" valign=\"middle\">";
  	  if (sessobj->sortcolumn == column)
  	    MAIN+=sortarrow;
  	  else
--- 8401,8407 ----
  	  MAIN+="<font class=\"to\" color=\"&imho_thcolor;\">"+MSG(M_TO)+"</font></th>";
  	  break;
  	case "subject":
! 	  MAIN+="<th class=\"subject\" bgcolor=\"&imho_thbgcolor;\" width=\"50%\" valign=\"middle\">";
  	  if (sessobj->sortcolumn == column)
  	    MAIN+=sortarrow;
  	  else
***************
*** 8168,8174 ****
  	    break;
  	  case "date":
  	    // Date
! 	    MAIN+="<td class=\""+(seen?"h":"")+"date\" bgcolor=\"&imho_tdbgcolor;\">"+(seen?"<b>":"")+parse_date(mail->imap->INTERNALDATE)+(seen?"</b>":"")+"</td>";
  	    break;
  	  case "from":
  	    // From
--- 8508,8514 ----
  	    break;
  	  case "date":
  	    // Date
! 	    MAIN+="<td class=\""+(seen?"h":"")+"date\" bgcolor=\"&imho_tdbgcolor;\">"+(seen?"<font color=red>":"")+parse_date(mail->imap->INTERNALDATE)+(seen?"<r>":"")+"</font></td>";
  	    break;
  	  case "from":
  	    // From
***************
*** 8179,8185 ****
  	      from_addr=replace(from_addr,"@.MISSING-HOST-NAME.","");
  	    } else
  	      from_name="-";
! 	    MAIN+="<td class=\""+(seen?"h":"")+"from\" bgcolor=\"&imho_tdbgcolor;\">"+(seen?"<b>":"")+fix_header(from_name?from_name:from_addr)+(seen?"</b>":"")+"</td>";
  	    break;
  	  case "to":
  	    // To
--- 8519,8525 ----
  	      from_addr=replace(from_addr,"@.MISSING-HOST-NAME.","");
  	    } else
  	      from_name="-";
! 	    MAIN+="<td class=\""+(seen?"h":"")+"from\" bgcolor=\"&imho_tdbgcolor;\">"+(seen?"<font color=red>":"")+fix_header(from_name?from_name:from_addr)+(seen?"</font>":"")+"</td>";
  	    break;
  	  case "to":
  	    // To
***************
*** 8192,8212 ****
  	      to_no = sizeof(mail->imap->ENVELOPE[TO_IDX]);
  	    } else
  	      to_name="-";
! 	    MAIN+="<td class=\""+(seen?"h":"")+"to\" bgcolor=\"&imho_tdbgcolor;\">"+(seen?"<b>":"")+fix_header(to_name?to_name:to_addr)+(to_no>1?" ...":"")+(seen?"</b>":"")+"</td>";
  	    break;
  	  case "subject":
  	    // Subject
! 	    MAIN+="<td class=\""+(seen?"h":"")+"subject\" bgcolor=\"&imho_tdbgcolor;\">"+(seen?"<b>":"")+"<a target=\""+id->misc->imho->sidetarget+"\" href=\""+id->misc->imho->nextsidetarget+"?actionread=1&msguid="+(string)sessobj->mails[i]->imap->UID+"&mbox="+http_encode_url(sessobj->mailbox[MB_FOLDERNAME_IDX])+"\">";
  	    string sub = fix_header(mail->imap->ENVELOPE[SUBJECT_IDX]);
  	    if (strlen(sub) == 0)
  	      MAIN+= "?";
  	    else
  	      MAIN+= sub;
! 	    MAIN+= "</a>"+(seen?"<b>":"")+"</td>";
  	    break;
  	  case "size":
  	    // Size
! 	    MAIN+="<td class=\""+(seen?"h":"")+"size\" bgcolor=\"&imho_tdbgcolor;\">"+(seen?"<b>":"")+((string) (((int) mail->imap["RFC822.SIZE"])/1024))+" K"+(seen?"</b>":"")+"</td>";
  	    break;
  	  }
  	MAIN+="</tr>\n";
--- 8532,8552 ----
  	      to_no = sizeof(mail->imap->ENVELOPE[TO_IDX]);
  	    } else
  	      to_name="-";
! 	    MAIN+="<td class=\""+(seen?"h":"")+"to\" bgcolor=\"&imho_tdbgcolor;\">"+(seen?"<font color=red>":"")+fix_header(to_name?to_name:to_addr)+(to_no>1?" ...":"")+(seen?"</font>":"")+"</td>";
  	    break;
  	  case "subject":
  	    // Subject
! 	    MAIN+="<td class=\""+(seen?"h":"")+"subject\" bgcolor=\"&imho_tdbgcolor;\">"+(seen?"<r>":"")+"<a target=\""+id->misc->imho->sidetarget+"\" href=\""+id->misc->imho->nextsidetarget+"?actionread=1&msguid="+(string)sessobj->mails[i]->imap->UID+"&mbox="+http_encode_url(sessobj->mailbox[MB_FOLDERNAME_IDX])+"\">";
  	    string sub = fix_header(mail->imap->ENVELOPE[SUBJECT_IDX]);
  	    if (strlen(sub) == 0)
  	      MAIN+= "?";
  	    else
  	      MAIN+= sub;
! 	    MAIN+= "</a>"+(seen?"<r>":"")+"</td>";
  	    break;
  	  case "size":
  	    // Size
! 	    MAIN+="<td class=\""+(seen?"h":"")+"size\" bgcolor=\"&imho_tdbgcolor;\">"+(seen?"<r>":"")+((string) (((int) mail->imap["RFC822.SIZE"])/1024))+" K"+(seen?"<r>":"")+"</td>";
  	    break;
  	  }
  	MAIN+="</tr>\n";
***************
*** 8484,8490 ****


      string buttons = "";
!     buttons+="<table border=\"0\" cellpadding=\"0\" cellspacing=\"0\"><tr><td><form target=\""+id->misc->imho->uptarget+"\" method=\"post\" action=\""+id->misc->imho->nextuptarget+"\">";
      // Reply
      buttons+="<imho_submit name=\"actionreply\" value=\""+MSG(M_REPLY)+"\" />";
      // Reply to all
--- 8824,8830 ----


      string buttons = "";
!     buttons+="<BR><table border=\"0\"  width=\"100%\" bgcolor=\"#a6d6d4\"  cellpadding=\"3\" cellspacing=\"1\"><tr><td bgcolor=\"#a6d6d4\"><form target=\""+id->misc->imho->uptarget+"\" method=\"post\" action=\""+id->misc->imho->nextuptarget+"\">";
      // Reply
      buttons+="<imho_submit name=\"actionreply\" value=\""+MSG(M_REPLY)+"\" />";
      // Reply to all
***************
*** 8493,8499 ****
      buttons+=" <imho_submit name=\"actionforward\" value=\""+MSG(M_FORWARD)+"\" />";
      buttons+="<input type=\"hidden\" name=\"mbox\" value="+html_encode_tag_value(sessobj->mailbox[MB_FOLDERNAME_IDX])+" />";
      buttons+="<input type=\"hidden\" name=\"msg"+sessobj->cmailuid+"\" value=\"1\" />";
!     buttons+="</form></td></tr><tr><td>";
      buttons+="<form method=\"post\" action=\""+id->misc->imho->nextpage+"\">";
      buttons+="<input type=\"hidden\" name=\"mbox\" value="+html_encode_tag_value(sessobj->mailbox[MB_FOLDERNAME_IDX])+" />";
      buttons+="<input type=\"hidden\" name=\"msg"+sessobj->cmailuid+"\" value=\"1\" />";
--- 8833,8839 ----
      buttons+=" <imho_submit name=\"actionforward\" value=\""+MSG(M_FORWARD)+"\" />";
      buttons+="<input type=\"hidden\" name=\"mbox\" value="+html_encode_tag_value(sessobj->mailbox[MB_FOLDERNAME_IDX])+" />";
      buttons+="<input type=\"hidden\" name=\"msg"+sessobj->cmailuid+"\" value=\"1\" />";
!     buttons+="</form></td></tr><tr><td bgcolor=\"#a6d6d4\">";
      buttons+="<form method=\"post\" action=\""+id->misc->imho->nextpage+"\">";
      buttons+="<input type=\"hidden\" name=\"mbox\" value="+html_encode_tag_value(sessobj->mailbox[MB_FOLDERNAME_IDX])+" />";
      buttons+="<input type=\"hidden\" name=\"msg"+sessobj->cmailuid+"\" value=\"1\" />";
***************
*** 8521,8531 ****
  	buttons+=" <imho_submit name=\"actionshowheaders\" value=\""+MSG(M_SHOWFULLHEADERS)+"\" /> ";
      }
      // Read previous/next
      if (prevuid!=-1)
!       buttons+="<nobr><imho_submit name=\"actionreadprev\" value=\"<- "+MSG(M_READPREV)+"\" />";
      if (nextuid!=-1)
!       buttons+="<imho_submit name=\"actionreadnext\" value=\""+MSG(M_READNEXT)+" ->\" /></nobr>";
!     buttons+="</form></td></tr></table>\n";

      int feature_addressbook=feature(FEAT_EDITADDRESSBOOK);

--- 8861,8881 ----
  	buttons+=" <imho_submit name=\"actionshowheaders\" value=\""+MSG(M_SHOWFULLHEADERS)+"\" /> ";
      }
      // Read previous/next
+     // AJT 15Aug2000 switch function depending on sort order
+     // remove arrow/pointers 15Sep2000
      if (prevuid!=-1)
!       if(sessobj->sessionsortorder == "forward")
!       buttons+="<nobr><imho_submit name=\"actionreadprev\" value=\""+MSG(M_READPREV)+"\" />";
!       else
!       buttons+="<nobr><imho_submit name=\"actionreadprev\" value=\""+MSG(M_READNEXT)+"\" />";
! /*      buttons+="<nobr><imho_submit name=\"actionreadprev\" value=\"<- "+MSG(M_READNEXT)+"\" />";*/
      if (nextuid!=-1)
!       if(sessobj->sessionsortorder == "forward")
!       buttons+="<imho_submit name=\"actionreadnext\" value=\""+MSG(M_READNEXT)+"\" />";
!       else
!       buttons+="<imho_submit name=\"actionreadnext\" value=\""+MSG(M_READPREV)+"\" />";
! /*      buttons+="<imho_submit name=\"actionreadnext\" value=\""+MSG(M_READPREV)+" ->\" />";*/
!     buttons+="</form></td></tr></table><BR>\n";

      int feature_addressbook=feature(FEAT_EDITADDRESSBOOK);

***************
*** 8565,8575 ****
      MAIN+="<br /><b>"+MSG(M_TIME)+":</b> "+fix_header(mail->headers->date);
      MAIN+="<br /><b>"+MSG(M_SUBJECT)+":</b> "+fix_header(mail->headers->subject);

      if (sessobj->showheaders) {
        foreach(indices(mail->headers), string h) {
! 	if (h != "to" && h != "from" && h != "cc" && h!="date" && h!="subject")
! 	  MAIN+="<br /><b>"+h+":</b> "+
! 	    fix_header(mail->headers[h]);
        }
      }

--- 8915,8928 ----
      MAIN+="<br /><b>"+MSG(M_TIME)+":</b> "+fix_header(mail->headers->date);
      MAIN+="<br /><b>"+MSG(M_SUBJECT)+":</b> "+fix_header(mail->headers->subject);

+ // patch from stewa@lysator.liu.se Sun Jul  2 20:41:16 2000
+ // uppercase by AJT 15Aug20000
      if (sessobj->showheaders) {
        foreach(indices(mail->headers), string h) {
! 	if (!(<"to","from","cc","date","subject">)[h])
! 	  foreach(mail->headers[h]/"\0",string h2)
! 	    MAIN+="<br /><b>"+String.capitalize(h)+":</b> "+fix_header(h2);
! /*	    MAIN+="<br /><b>"+h+":</b> "+fix_header(h2);*/
        }
      }

***************
*** 8605,8613 ****
  	}
        } else {
  	string name=fix_coding(mess->disp_params["filename"] || mess->params["name"] || "");
! 	MAIN+="<table width=100% bgcolor=&imho_tdbgcolor;><tr><td>";

  	int linkshown = 0;
  	if(!(!msgshown && mess->type == "text" &&
  	   (parenttype->type == "multipart" &&
  	    parenttype->subtype == "alternative")?((feature(FEAT_SHOWHTML)&&(sessobj->showhtml=="yes"))?0:mess->subtype=="plain"):(mess->subtype == "plain"))) {
--- 8958,8985 ----
  	}
        } else {
  	string name=fix_coding(mess->disp_params["filename"] || mess->params["name"] || "");
! 	MAIN+="<br><br><table width=100% bgcolor=&imho_tdbgcolor;><tr><td>";

  	int linkshown = 0;
+ // AJT special for message/rfc822 treat as text/plain... 17Aug2000
+ // it even seems to work...
+ 	if(mess->type =="message" && mess->subtype == "rfc822") {
+ 	  linkshown = 1;
+ 	  MAIN+="<a href=\""+id->misc->imho->nextpage+"/"+(sizeof(name)?http_encode_url(name):"unknown")+"?mailpart="+partno+"\"><img src=\""+image_from_type(((mess->type)/"/")[0])+"\" alt=\"\" border=\"0\"> ";
+ 	  MAIN+=replace(MSGA(M_ATTACHMENTLINK,({sizeof(name)?html_encode_string(name)+", ":"","\0" })),"\0",mess->type+"/"+mess->subtype )+"</a>";
+ 	  string data = mess->getdata()||"";
+ 	  msgshown = 1;
+ 	    mixed err=0;
+ 	    string _mess;
+ 	    err=catch { _mess=Locale.Charset->decoder(mess->charset)->feed(data)->drain(); };
+ 	    if(err)
+ 	      MAIN+=" "+replace(MSGA(M_CHARSETWARNING,({ "\0" })),"\0",html_encode_string(mess->charset))+"<hr />";
+ 	      MAIN+="<BR><font face=\"Courier\">";
+ 	      MAIN+=replace(make_links(id, fixstring(err?data:_mess)),
+ 			    "  ", "&nbsp;&nbsp;");
+ 	      MAIN+="</font>";
+ 	}
+ // end of message/rfc822
  	if(!(!msgshown && mess->type == "text" &&
  	   (parenttype->type == "multipart" &&
  	    parenttype->subtype == "alternative")?((feature(FEAT_SHOWHTML)&&(sessobj->showhtml=="yes"))?0:mess->subtype=="plain"):(mess->subtype == "plain"))) {
***************
*** 8649,8654 ****
--- 9021,9033 ----
  	     (!mess->headers["content-id"] ||
  	      !refparts[mess->headers["content-id"]]))
  	    MAIN+="<hr /><img src=\""+id->misc->imho->nextpage+"/"+http_encode_url(name)+"?mailpart="+partno+"\" />";
+ //make text/plain attachments have a link 22Oct2000,23Oct2000
+ 	if(mess->type == "text" && mess->subtype == "plain" ) {
+ 	  MAIN+="<br><br><a href=\""+id->misc->imho->nextpage+"/"+(sizeof(name)?http_encode_url(name):"mail"+partno+".text")+"?mailpart="+partno+"\"> ";
+ 	  MAIN+="right click here to download text part </a>";
+ /*	  MAIN+=replace(MSGA(M_ATTACHMENTLINK,({sizeof(name)?html_encode_string(name)+", ":"","\0" })),"\0",mess->type+"/"+mess->subtype )+"</a>";*/
+ 	  MAIN+="<br>";
+ 	}

  	MAIN+="</td></tr>\n</table><br />\n";
  	part++;
***************
*** 9089,9094 ****
--- 9468,9476 ----
      return "<b>Could not log in:</b> Wrong login or password. Try again. ";
    case M_IMAPERROR:
      return "There was a problem talking to the mail (IMAP) server. Please contact the administrator.";
+ /* AJT 10Oct2000 */
+   case M_NOCOOKIE:
+     return "<font color=\"red\">Your browser does not support <a href=\"http://webmail.liv.ac.uk/cookies.html\">cookies</a></font>"; //22Oct2000 AJT
    case M_SMTPERROR:
      return "There was a problem talking to the outgoing mail (SMTP) server, and the mail was not sent. The recipient of the mail might not exist.";
    case M_SAVEDUSERINTERFACE:
***************
*** 9335,9341 ****
      return "Suggestions";

    case M_LOGOUTMSG:
!     return "<b>You are logged out.</b> Close this browser window to make sure no one can read your mail.<br /><br />Reload this page to login.";
    case M_LOGGEDOUT:
      return "Logged out";

--- 9717,9723 ----
      return "Suggestions";

    case M_LOGOUTMSG:
!     return "<BR><BR><b>You are logged out.</b><BR><BR><font color =red> Close this browser window to make sure no one can read your mail.</font><br /><br />Reload this page to login.";
    case M_LOGGEDOUT:
      return "Logged out";

***************
*** 9453,9459 ****
    case M_SEARCHAND:
      return("AND");
    case M_SEARCHSHOWINGMAIL:
!     return("Showing mail with ");

    case M_YES:
      return "Yes";
--- 9835,9843 ----
    case M_SEARCHAND:
      return("AND");
    case M_SEARCHSHOWINGMAIL:
!     return("(clear search by click on check for new mail) Showing mail with ");
! //AJT 01Oct2000 more help for search
! //    return("Showing mail with ");

    case M_YES:
      return "Yes";




